src/analyses/variable-sensitivity/dependency_context_abstract_object.cpp

```diff
@@ -0,0 +1,397 @@
+
+\*******************************************************************/
+abstract_objectt::locationst
+  dependency_context_abstract_objectt::get_last_written_locations() const
+{
+  return last_written_locations;
+}
+
+/*******************************************************************\
+
+Function: dependency_context_abstract_objectt::read
+
+  Inputs:
+   env - the abstract environment
+   specifier - a modifier expression, such as an array index or field specifier
+          used to indicate access to a specific component
+
+ Outputs: The abstract_objectt representing the value of that component. For
+          this default implementation, we just return `this`. Sub-classes are
+          expected to extend this implementation to match their behaviour.
+
+ Purpose: A helper function to evaluate an abstract object contained
+          within a container object. More precise abstractions may override this
+          to return more precise results.
+
+\*******************************************************************/
+abstract_object_pointert dependency_context_abstract_objectt::read(
+  const abstract_environmentt &env,
+  const exprt &specifier,
+  const namespacet &ns) const
+{
+  return ao->read(env, specifier, ns);
+}
+
+/*******************************************************************\
+
+Function: dependency_context_abstract_objectt::write
+
+  Inputs:
+   environment - the abstract environment
+   stack - the remaining stack of expressions on the LHS to evaluate
+   specifier - the expression uses to access a specific component
+   value - the value we are trying to write to the component
+
+ Outputs: The abstract_objectt representing the result of writing
+          to a specific component.
+
+ Purpose: A helper function to evaluate writing to a component of an
+          abstract object. More precise abstractions may override this to
+          update what they are storing for a specific component.
+
+\*******************************************************************/
+abstract_object_pointert dependency_context_abstract_objectt::write(
+  abstract_environmentt &environment,
+  const namespacet &ns,
+  const std::stack<exprt> stack,
+  const exprt &specifier,
+  const abstract_object_pointert value,
+  bool merging_write) const
+{
+  // Need to ensure the result of the write is still wrapped in a dependency
+  // context
+  const auto &result=
+    std::dynamic_pointer_cast<dependency_context_abstract_objectt>(
+      mutable_clone());
+
+  // But delegate the write to the child
+  abstract_object_pointert updated_child=
+    ao->write(environment, ns, stack, specifier, value, merging_write);
+
+  result->set_child(updated_child);
+
+  // This is a bit horrible, but not sure how to express this
+  // in a more general pattern without also pushing knowledge
+  // of last_written_location housekeeping into the other
+  // *_abstract_objectt classes.
+  if(std::dynamic_pointer_cast<const full_struct_abstract_objectt>
```

Comment:
Could you extend this comment to cover what this is doing? Is this making so when I do `a.x = 4`, the abstract object for `a` last written location being set to the same as `a.x`?

Reply:
I think it might be preferable to push hose keeping into the other objects as this looks likely to trip over someone implementing a new `struct_abstract_objectt`.